core/capabilities/webapi/target/connector_handler.go

```diff
@@ -0,0 +1,115 @@
+package target
+
+import (
+	"context"
+	"encoding/json"
+	"sort"
+	"sync"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/api"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/connector"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers/common"
+	"github.com/smartcontractkit/chainlink/v2/core/services/gateway/handlers/webcapabilities"
+)
+
+var _ connector.GatewayConnectorHandler = &ConnectorHandler{}
+
+type ConnectorHandler struct {
+	gc            connector.GatewayConnector
+	lggr          logger.Logger
+	responseChs   map[string]chan *api.Message
+	responseChsMu sync.Mutex
+	rateLimiter   *common.RateLimiter
+}
+
+func NewConnectorHandler(gc connector.GatewayConnector, config Config, lgger logger.Logger) (*ConnectorHandler, error) {
+	rateLimiter, err := common.NewRateLimiter(config.RateLimiter)
+	if err != nil {
+		return nil, err
+	}
+	responseChs := make(map[string]chan *api.Message)
+	return &ConnectorHandler{
+		gc:            gc,
+		responseChs:   responseChs,
+		responseChsMu: sync.Mutex{},
+		rateLimiter:   rateLimiter,
+		lggr:          lgger,
+	}, nil
+}
+
+// HandleSingleNodeRequest sends a request to first available gateway node and blocks until response is received
+// TODO: handle retries and timeouts
+func (c *ConnectorHandler) HandleSingleNodeRequest(ctx context.Context, messageID string, payload []byte) (*api.Message, error) {
+	ch := make(chan *api.Message, 1)
+	c.responseChsMu.Lock()
+	c.responseChs[messageID] = ch
+	c.responseChsMu.Unlock()
+	l := logger.With(c.lggr, "messageID", messageID)
+	l.Debugw("sending request to gateway")
+
+	body := &api.MessageBody{
+		MessageId: messageID,
+		DonId:     c.gc.DonID(),
+		Method:    webcapabilities.MethodWebAPITarget,
+		Payload:   payload,
+	}
+
+	// simply, send request to first available gateway node from sorted list
+	// this allows for deterministic selection of gateay node receiver for easier debugging
+	gatewayIDs := c.gc.GatewayIDs()
+	sort.Strings(gatewayIDs)
+
+	err := c.gc.SignAndSendToGateway(ctx, gatewayIDs[0], body)
```

Comment:
is `gatewayIDs` guaranteed to have length?

Reply:
sounds like a fatal set up error if it doesn't. But it doesn't seem like we have that validation anywhere. I can handle it more gracefully here.